### PR TITLE
Fix other escape strings from JS

### LINKED ISSUE #75: semantic: string literal escape sequence handing
Discovered as part of the https://github.com/bblfsh/bblfshd/issues/268#issuecomment-475046348

This is an ☂️ issue for multiple driver failures \w `invalid syntax`.

E.g driver fails to parse e.g [this file](https://ckeditor4-preview.ckeditor.com/next/tests/core/htmlparser/fragment/fromhtml.js) with 
```
     check: key "value": invalid syntax
```

Normalization mapping for semantic mode needs to be updated.

 - [x] `check: key "value": invalid syntax ("\0\0\0\0\0\0\0\0")`  #79 
           double quoted literals need to have same js escape sequence handled as #64
 - [ ] `check: key "value": invalid syntax ("\.\.")"`
 - [ ] ...

132 file examples could be obtained using

```
grep "language=javascript$" bblfshd.log | grep -oP "rpc error: code = \K(.*)" | grep "invalid syntax" | sort | less
```

### PR DESCRIPTION
**Pull request by bzz**
Fixes the rest of the #75  - case of `check: key "value": invalid syntax ("\ ")"`

Details of the applied solution are described in https://github.com/bblfsh/javascript-driver/pull/81#issuecomment-483262838

### CODE REVIEW

File: driver/normalizer/strconv_test.go [hunk 1]
```diff
@@ -26,10 +26,22 @@ var testCasesUnquoteSingle = []struct {
 	{`'\0\0'`, "\u0000\u0000", "'\\x00\\x00'"},
 }
 
+var testCasesUnquoteDouble = []struct {
+	quoted   string
+	unquoted string
+}{
+	// {`"\0\0\0"`, "\x00\x00\x00"},
```
Comment (creachadair) [inline driver/normalizer/strconv_test.go]: Is this a TODO, or is it dead?

File: fixtures/string-literal.js [hunk 2]
```diff
@@ -9,3 +9,5 @@ var escComma = '\0COMMA'+Math.random()+'\0';
 var escPeriod = '\0PERIOD'+Math.random()+'\0';
 
 d = "\0\0\0\0\0\0\0\0"
+
+c = "\.\1 or really \anthining"
```
Comment (creachadair) [inline fixtures/string-literal.js]:
```suggestion
c = "\.\1 or really \anything"
```

**Comment by bzz** [comment]
Rebased on laster master and early feedback amended.

Right now it's still WIP as it only contains the reproducible example + tests but no a fix yet.

**Comment by bzz** [comment]
As we now have discovered many edge cases where escape sequence handing in the native driver is very different from Go - current approach from #64  does not seem to scale well, so a proposed solution that we discussed with @dennwc yesterday is:
 - in *annotated* mode: always take un-normalized value as token (includes `\"`, `\'`, etc)
 - in *semantic* mode, always take normalized value as toke (does not include such characters)

That would allow us to avoid re-implementing complex JS escape sequence handling on the Go side. 

Although this seems to be consistent with initial intent https://github.com/bblfsh/javascript-driver/issues/32#issuecomment-451142636 and seems to be generally the case for other drivers, I'm not sure how that affects clients expectations (e.g see #32).

**Comment by creachadair** [comment]
> * in _native/annotated_ mode: always take un-normalized value as token (includes `\"`, `\'`, etc)
> * in _semantic_ mode, always take normalized value as toke (does not include such characters)
> 
> That would allow us to avoid re-implementing complex JS escape sequence handling on the Go side.

That seems reasonable to me. I also like about it that its easy to understand which to expect: If you're upstream of canonicalization, you get the ambient encoding, downstream you get the normalized form.

Does this mean, though, that we will have to explicitly define the string encoding rules we use inside the canonical AST? (I didn't find it anywhere, and it looks like right now we implicitly assume Go's rules).

**Comment by bzz** [comment]
I guess that is something I'm trying to understand here as well.

After a series of experiments, I learned go, java, bash, cpp, csharp - are the drivers where the original quotas used for literals are preserved in _annotated mode_.

php and python do not preserve that information and thus `'b\ncq'` shows up as `"@token":"b\ncq","@type":"Str"`

**Comment by bzz** [comment]
The format below is: 

`string literal`
and it's represenation in
```
annotated
semantc
```
modes.

--------------

### go 
`"b\nc"`

```
"@token":"\"b\\nc\""
"@type":"uast:String","Format":"","Value":"b\nc"
```
 
### java
`"b\nc"`

```
"@token":"\"b\\nc\""
"@type":"uast:String","Format":"","Value":"b\nc"
 ```

### javascript
`"b\nc"`

```
"@token":"\"b\\nc\""
"@type":"uast:String","Format":"","Value":"b\nc"
```

`'b\ncq'`
```
"@token":"'b\\ncq'"
"@type":"uast:String","Format":"single","Value":"b\ncq"
```

### bash
`"b\nc"`

```
"@token":"\"b\\nc\"","@type":"string","children":["@role":["Expression","String","Literal"],"@token":"b\\nc","@type":"string_content"]
"@type":"uast:String","Format":"","Value":"b\\nc"
```

`'b\ncq'`
```
"@token":"'b\\ncq'","@type":"unevaluated_string2","children":[]
"@type":"uast:String","Format":"","Value":"'b\\ncq'"
 ```

### cpp
`"b\nc"` (not sure yet, why is it twice :/ steps to reproduce in details)
```
"@token":"\"b\\nc\"","@type":"CPPASTLiteralExpression","ExpressionType":"const char [4]","ExpressionValueCategory":"LVALUE","IsLValue":true,"kind":"string_literal"
"@type":"uast:String","Format":"","Value":"b\nc"

"@token":"\"b\\nc\"","@type":"CPPASTLiteralExpression","ExpressionType":"const char [4]","ExpressionValueCategory":"LVALUE","IsLValue":true,"kind":"string_literal"
"@type":"uast:String","Format":"","Value":"b\nc"
```

<details>

```cpp
const char* b = 'b';
const char* bc = "b\nc";
```

```
 bblfsh-cli -o json -l cpp -m annotated \
    -q "//*[@role='String' and @role='Literal']" cpp.cpp \
    | jq -c '.[] | del(."@pos") | del(.Token."@pos") | del(."@role")'

 bblfsh-cli -o json -l cpp -m semantic \
    -q '//uast:String' cpp.cpp \
   | jq -c '.[] | del(."@pos")'
```

</details>
 
### csharp
`"b\nc"`

```
"@type":"StringLiteralExpression","IsMissing":false,"IsStructuredTrivia":false,"Token":{"@role":["Literal","String"],"@type":"StringLiteralToken","IsMissing":false,"Text":"\"b\\nc\"","Value":"b\nc","ValueText":"b\nc"}

"@type":"uast:String","Format":"","Value":"b\nc"
```
 
### php
`"b\nc"`
```
"@token":"b\nc","@type":"Scalar_String","attributes":{"kind":2}
"@type":"uast:String","Format":"","Value":"b\nc"
```

`'b\ncq'`
```
"@token":"b\\ncq","@type":"Scalar_String","attributes":{"kind":1}
"@type":"uast:String","Format":"raw","Value":"b\\ncq"
```

### python
`"b\nc"`
```
"@token":"b\nc","@type":"Str"
"@type":"uast:String","Format":"","Value":"b\nc"
```

`'b\ncq'`
```
"@token":"b\ncq","@type":"Str"
"@type":"uast:String","Format":"","Value":"b\ncq"
```

**Comment by dennwc** [comment]
- Semantic in Bash is incorrect (should be `"b\ncq"`)
- Annotated in PHP and Python is incorrect (should be `\"b\\ncq\"` and `'b\\ncq'`)

Everything else looks good.

In Annotated the string should match the source file (including quotes). In Semantic there should be no quotes and the string should be unescaped according to the language rules (same as if you print it from that language).

**Comment by bzz** [comment]
I'm using this as kind-of umbrella issue for these type of problems, so will create new appropriate issues for Bash, PHP and Python.

So, the same approach will be used as solution for https://github.com/bblfsh/java-driver/issues/111 and this one:
 - keep annotated mode as it is now
 - in semantic mode, instead of unquoting annotated one on Go side (leads to these bugs), we'll just use  a different AST node, that already provides a normalized string, according the native language rules.

**Comment by bzz** [comment]
Fix applied in 683324b : 
  - in Annotated, `rawValue` is now preserved as `normValue` (contains normalised string, previously discarded)
  - in Semantic, `format: single` is kept the same way but `normValue` is used instead of manual unquoting

All fixtures updated in fbed224. 
@dennwc @creachadair ci is great and it's ready for review now.

File: driver/normalizer/normalizer.go [hunk 3]
```diff
@@ -136,7 +135,8 @@ var Normalizers = []Mapping{
 	)),
 	MapSemantic("StringLiteral", uast.String{}, MapObj(
 		Fields{
-			{Name: "value", Op: singleQuote{Var("val")}},
+			{Name: "normValue", Op: Var("val")},
+			{Name: "value", Drop: true, Op: singleQuote{Any()}}, // only used in Annotated
```
Comment (dennwc) [inline driver/normalizer/normalizer.go]:
```suggestion
			{Name: "value", Drop: true, Op: Any()}, // only used in Annotated
```
No need to unquote if you drop it anyway.

Comment (bzz) [reply driver/normalizer/normalizer.go]: At first, I did exactly this BUT then we'll loose the `format: single` inside the Semantic node.

So I have [changed the impl of `singleQuote`](https://github.com/bblfsh/javascript-driver/pull/81/files#diff-da412cf66633e782d4653e9a0a1b54bdR380) to become no-op that just _detects_ the quotation and [matches those literals first](https://github.com/bblfsh/javascript-driver/pull/81/files#diff-da412cf66633e782d4653e9a0a1b54bdR143) - that way we can still use this signal.

Comment (dennwc) [reply driver/normalizer/normalizer.go]: Sorry, missed that part. Makes total sense now, but it needs a comment.

As an alternative, you can drop the `Construct` method, remove the sub-operation from it and use this custom transform as a filter: `Check(singleQuote{}, Any())`.

Comment (bzz) [reply driver/normalizer/normalizer.go]: Addressed in fc57c535f68cff874cad98a0bcd3088c7dc6d200 - could you please check if I got that right?

**Comment by bzz** [comment]
All feedback addressed in fc57c535f68cff874cad98a0bcd3088c7dc6d200 - could you please give it a final glance, @dennwc ?

File: driver/normalizer/normalizer.go [hunk 4]
```diff
@@ -378,17 +377,5 @@ func (op singleQuote) Check(st *State, n nodes.Node) (bool, error) {
...
-	n, err := op.op.Construct(st, n)
-	if err != nil {
-		return nil, err
-	}
-	sn, ok := n.(nodes.String)
-	if !ok {
-		return nil, ErrUnexpectedType.New(nodes.String(""), n)
-	}
-	return sn, nil
+	return true, nil
```
Comment (dennwc) [inline driver/normalizer/normalizer.go]: Could be
```golang
return strings.HasPrefix(s, `'`) && strings.HasSuffix(s, `'`), nil
```
(optional)

Comment (bzz) [reply driver/normalizer/normalizer.go]: So true! Sorry for missing that out, addressed in f20fe64519b06f4bb2d9efba44d6898c4f7d0451 \w history being re-written.

**Comment by bzz** [comment]
Thank you every one for reviews! Merging after CI is green
